.circleci/config.yml

```diff
@@ -1113,6 +1114,77 @@ commands:
           background: true
 
 jobs:
+  # Allows jobs in the "nightly" workflow to run if either of the following
+  # conditions are true:
+  #
+  # 1. A scheduled pipeline is run with the "nightly_tests" trigger name.
+  #
+  #    See "https://circleci.com/docs/schedule-pipelines-with-multiple-workflows/".
+  #
+  #    The CircleCI Admin will set up a trigger for when the nightly scheduled
+  #    pipeline is to be run. Typically, the nightly scheduled pipeline will
+  #    only be configured to run on the "main" branch or a release branch.
+  #
+  #    See "https://circleci.com/docs/set-a-nightly-scheduled-pipeline/".
+  #
+  #    The "pipeline.trigger_source" and "pipeline.schedule.name" are built-in
+  #    pipeline values and, therefore, do not need to be declared in the
+  #    "parameters:" section.
+  #
+  #    See "https://circleci.com/docs/variables/#pipeline-values".
+  #
+  # 2. The developer requests that the nightly tests be run on demand at the
+  #    gate by adding the 'ci-run-nightly' label to their pull request.
+  #    pull request.
+  #
+  # If neither condition is true, the "nightly" workflow is cancelled.
+  #
+  # For this mechanism to work, all jobs in the "nightly" workflow must have
+  # a "requires" statement listing "check-if-nightly-tests-should-run".
+  check-if-nightly-tests-should-run:
+    docker:
+      - image: <<pipeline.parameters.docker-image>>
+    steps:
+      # Install the GitHub CLI. Requires the 'gh' orb.
+      - gh/install
+      - checkout
+      - run:
+          name: Check if nightly tests should run
+          command: |
+            # Check if the nightly workflow was triggered by the scheduled
+            # pipeline.
+            if [ "scheduled_pipeline" == "<<pipeline.trigger_source>>" ] &&
+               [ "nightly_tests" == "<<pipeline.schedule.name>>" ]
+            then
+                echo "Nightly workflow triggered from scheduled pipeline"
+                circleci-agent step halt
+                exit 0
+            fi
+
+            # Check if the 'ci-run-nightly' label was set on the pull request,
+            # which indicates that we want the nightly tests to run on demand.
+            if gh pr view | grep "^labels:" | grep -w ci-run-nightly
+            then
+                echo "Nightly workflow triggered from 'ci-run-nightly' label on pull request"
+                circleci-agent step halt
+                exit 0
+            fi
+
+            echo "Canceling the nightly workflow, as it was not requested from either a scheduled pipeline or 'ci-run-nightly' label."
```

Comment:
So we will see a lot of cancelled nightly workflow instances? Will this message be easily findable to explain the situation?

Reply:
Every pull request will have a `Nightly` workflow.  It will show up as `Canceled` if you did not include the `ci-run-nightly` label.  You can see what it looks like in `Test Case #3`.